### PR TITLE
[Diagnostics] Handle CoerceExpr conversion failure in contextual mismatch

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
While working in #27895, more specifically in https://github.com/apple/swift/pull/27895/commits/82529a6e7db2c46b86d8ede9bb520cfc112f8a6c I note that the cannot convert diagnostic for expr `Double(1) as Double as String` is being emitted via fix but still going re-typechecking passing through `CSDiag`. Note that actually when tried to put a `if (Options.contains(ConstraintSystemFlags::SubExpressionDiagnostics))` in `applySolutionFixes`.
So this PR is to handle coercion conversion error early on `repairFailures` and avoid this. 
This also handles the TODO diagnostic in `test/type/subclass_composition.swift`.

cc @xedin @hborla :))

### CODE REVIEW

**Comment by xedin** [comment]
@LucianoPAlmeida Thank you, I'll take a look on Monday! Meanwhile, can check you check how much fails if you remove `FailureDiagnosis::visitCoerceExpr`?

**Comment by LucianoPAlmeida** [comment]
Sure @xedin, will check as soon as possible 👍

**Comment by LucianoPAlmeida** [comment]
@xedin removing `FailureDiagnosis::visitCoerceExpr` only fails 5 tests:
```
Failing Tests (5):
    Swift(macosx-x86_64) :: Generics/conditional_conformances_literals.swift
    Swift(macosx-x86_64) :: Sema/suppress-argument-labels-in-types.swift
    Swift(macosx-x86_64) :: expr/cast/metatype_casts.swift
    Swift(macosx-x86_64) :: expr/cast/optional.swift
    Swift(macosx-x86_64) :: type/subclass_composition_objc.swift

  Expected Passes    : 4950
  Expected Failures  : 23
  Unsupported Tests  : 122
  Unexpected Failures: 5
``` 
I do remove some obsolete code from it, but can try to figure out the missing cases in a follow-up PR so it can be removed for good \o/ 
Most of them are coercions involving metatypes, so probably handle a complex locator ...

**Comment by xedin** [comment]
@LucianoPAlmeida That would be great!

**Review by xedin — APPROVED** [review]
Looks great! I have left a minor naming suggesting inline.

File: lib/Sema/CSDiagnostics.cpp [hunk 1]
```diff
@@ -2220,6 +2223,28 @@ bool ContextualFailure::diagnoseMissingFunctionCall() const {
   return true;
 }
 
+bool ContextualFailure::diagnoseConversionInCoercion() const {
```
[the remark this answers is not in the document]
Comment (LucianoPAlmeida) [reply lib/Sema/CSDiagnostics.cpp]: Sure, Thanks @xedin :)

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by LucianoPAlmeida** [comment]
@xedin Thanks, I'm starting to work on handle the other failures, maybe we'll have something soon :)

**Comment by xedin** [comment]
Sounds good, thank you!

**Pull request merged**